src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs

```diff
@@ -391,6 +391,21 @@ private async Task<EntryComponent> ExecuteRequests(Hl7.Fhir.Model.Bundle respons
 
                         await request.Handler.Invoke(httpContext);
 
+                        // we will retry a 429 one time per request in the bundle
+                        if (httpContext.Response.StatusCode == (int)HttpStatusCode.TooManyRequests)
+                        {
+                            _logger.LogTrace("BundleHandler received 429 message, attempting retry.  HttpVerb:{httpVerb} BundleSize: {_requestCount} entryIndex:{entryIndex}", httpVerb, _requestCount, entryIndex);
+                            int retryDelay = 2;
+                            var retryAfterValues = httpContext.Response.Headers.GetCommaSeparatedValues("Retry-After");
+                            if (retryAfterValues != StringValues.Empty && int.TryParse(retryAfterValues[0], out var retryHeaderValue))
+                            {
+                                retryDelay = retryHeaderValue;
+                            }
+
```

Comment:
Is it worth to retry in absence of "Retry-After" header?

Feels that delay 2 milliseconds wouldn't do much to solve problem...

Reply:
I was worried that 2 seconds might be too long :).  I used 2 seconds as a fall back, because we have similar logic in the CosmosFhirStore code.  I would still like to try, given that we did receive a 429.

Reply:
[https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.delay?view=net-6.0](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.delay?view=net-6.0)

reates a cancellable task that completes after a specified number of *milliseconds*.

It's not a seconds...

Reply:
Thank for you for catching it Ivan!  Sorry I missed the "milliseconds" in your previous comment.  I merged this too quickly!  I will put up a new PR to fix it.